nomad/job_endpoint_validators.go

```diff
@@ -0,0 +1,70 @@
+package nomad
+
+import (
+	"strings"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/pkg/errors"
+)
+
+type jobNamespaceConstraintCheckHook struct {
+	srv *Server
+}
+
+func (jobNamespaceConstraintCheckHook) Name() string {
+	return "namespace-constraint-check"
+}
+
+func (c jobNamespaceConstraintCheckHook) Validate(job *structs.Job) (warnings []error, err error) {
+	// This was validated before and matches the WriteRequest namespace
+	ns, err := c.srv.State().NamespaceByName(nil, job.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	if ns == nil {
+		return nil, errors.Errorf("job %q is in nonexistent namespace %q", job.ID, job.Namespace)
+	}
+
+	var disallowedDrivers []string
+	for _, tg := range job.TaskGroups {
+		for _, t := range tg.Tasks {
+			if !taskValidateDriver(t, ns) {
+				disallowedDrivers = append(disallowedDrivers, t.Driver)
+			}
+		}
+	}
+	if len(disallowedDrivers) > 0 {
+		if len(disallowedDrivers) == 1 {
+			return nil, errors.Errorf(
+				"used task driver '%s' is not allowed in namespace %s", disallowedDrivers[0], ns.Name)
+
+		} else {
+			for i := 0; i < len(disallowedDrivers); i++ {
+				disallowedDrivers[i] = "'" + disallowedDrivers[i] + "'"
+			}
+			return nil, errors.Errorf(
+				"used task drivers %s are not allowed in namespace %s", strings.Join(disallowedDrivers, ", "), ns.Name)
```

Comment:
Likewise, we could either use `%q` here around the string slice or we could use some clever `strings.Join` to reduce the number of string operations we have to do: https://play.golang.com/p/u3ZsG6x-HAa

Either way works for me.